zulia-server/src/main/java/io/zulia/server/search/StatCombiner.java

```diff
@@ -26,30 +37,314 @@ public int getShardIndex() {
 		}
 	}
 
+	private static class FacetStatsWithShardIndex {
+		private final FacetStatsInternal facetStats;
+		private final int shardIndex;
+
+		public FacetStatsWithShardIndex(FacetStatsInternal facetStats, int shardIndex) {
+			this.facetStats = facetStats;
+			this.shardIndex = shardIndex;
+		}
+
+		public FacetStatsInternal getFacetStats() {
+			return facetStats;
+		}
+
+		public int getShardIndex() {
+			return shardIndex;
+		}
+	}
+
 	private final List<StatGroupWithShardIndex> statGroups;
-	private final int[] shardIndexes;
 	private final StatRequest statRequest;
 	private final int shardReponses;
 
 	public StatCombiner(StatRequest statRequest, int shardReponses) {
 		this.statRequest = statRequest;
 		this.shardReponses = shardReponses;
 		this.statGroups = new ArrayList<>(shardReponses);
-		this.shardIndexes = new int[shardReponses];
 	}
 
-	public void handleStatGroupForShard(StatGroup statGroup, int shardIndex) {
+	public void handleStatGroupForShard(StatGroupInternal statGroup, int shardIndex) {
 		statGroups.add(new StatGroupWithShardIndex(statGroup, shardIndex));
 	}
 
-	public StatGroup getCombinedStatGroup() {
+	public ZuliaQuery.StatGroup getCombinedStatGroupAndConvertToExternalType() {
+		// Get global stats
+		List<FacetStatsWithShardIndex> globalStatsInternal = new ArrayList<>();
+		statGroups.forEach(sg -> globalStatsInternal.add(new FacetStatsWithShardIndex(sg.getStatGroup().getGlobalStats(), sg.getShardIndex())));
+
+		// Create a map grouping of lists of facets to be merged
+		HashMap<String, List<FacetStatsWithShardIndex>> facetStatsGroups = new HashMap<>();
+		statGroups.forEach(statGroup ->
+				// Group all sets of facet stats
+				statGroup.getStatGroup().getFacetStatsList().forEach(facetStats -> {
+					String localName = facetStats.getFacet();
+					List<FacetStatsWithShardIndex> temp2 = facetStatsGroups.getOrDefault(localName, new ArrayList<>());
+					temp2.add(new FacetStatsWithShardIndex(facetStats, statGroup.getShardIndex()));
+					facetStatsGroups.put(localName, temp2);
+				}));
+
+		// Send each group through the converter to generate a list
+		FacetStats.Builder globalStats = convertAndCombineFacetStats(globalStatsInternal);
+		List<FacetStats.Builder> facetStats = facetStatsGroups.values().stream().map(this::convertAndCombineFacetStats).toList();
+
+		// Find / bound errors in the sum element
+		for (FacetStats.Builder fs : facetStats) {
+			if (fs.getHasError()) {
+				// The error is the summation across all shards of the smallest value returned by each shard which did not return this facet
+				List<Integer> missing = getNonReportingShards(facetStatsGroups.get(fs.getFacet()));
```

Comment:
this is a great and readable way to do this

Reply:
That section got re-written a bunch of times lol